transport/src/main/java/io/netty/channel/SingleThreadEventLoop.java

```diff
@@ -71,11 +81,104 @@ public ChannelFuture register(final Channel channel, final ChannelPromise promis
         return promise;
     }
 
+    /**
+     * Adds a task to be run once at the end of next (or current) {@code eventloop} iteration.
+     *
+     * @param task to be added.
+     */
+    public void onEventLoopIteration(Runnable task) {
+        ObjectUtil.checkNotNull(task, "task");
+        if (isShutdown() || isShuttingDown() || isTerminated()) {
+            reject();
+        }
+
+        tailTasks.add(task);
+
+        if (wakesUpForTask(task)) {
+            wakeup(inEventLoop());
+        }
+    }
+
+    /**
+     * Removes a task that was added previously via {@link #onEventLoopIteration(Runnable)}.
+     *
+     * @param task to be removed.
+     *
+     * @return {@code true} if the task was removed as a result of this call.
+     */
+    public boolean removeOnEventLoopIterationTask(Runnable task) {
+        return tailTasks.remove(ObjectUtil.checkNotNull(task, "task"));
+    }
+
     @Override
     protected boolean wakesUpForTask(Runnable task) {
         return !(task instanceof NonWakeupRunnable);
     }
 
+    @Override
+    protected boolean runAllTasks(long timeoutNanos) {
+        if (lastRunTimedOut) {
+            lastRunTimedOut = false;
+            Runnable task = pollTask();
+            if (task != null) {
+                long startNanos = nanoTime();
+                final long deadline = startNanos + timeoutNanos;
+                long runTasks = 0;
+                long lastExecutionTime;
+                for (;;) {
+                    try {
+                        task.run();
+                    } catch (Throwable t) {
+                        logger.warn("A task raised an exception.", t);
+                    }
+
+                    runTasks ++;
+
+                    // Check timeout every 64 tasks because nanoTime() is relatively expensive.
+                    // XXX: Hard-coded value - will make it configurable if it is really a problem.
+                    if ((runTasks & 0x3F) == 0) {
```

Comment:
Yeah, I wasn't happy about code copy but didn't want to touch too many things in the first iteration.

I have refactored the super-class code and reused it in `SingleThreadEventLoop`, PTAL.